### ISSUE TITLE
Adding new article on custom controls

### ISSUE THREAD

**Issue by mshwf** [MicrosoftDocs/xamarin-docs#1666]
I've created multiple controls in Xamarin due to the shortage of these controls in Xamarin or due to the incompetent solutions from 3rd party or just due to customer requirements, my experience with WPF played a major part of designing these controls, unfortunately when it comes to custom controls the Xamarin docs lay behind the scene, so I think I can bring this topic to the docs, 
for example this [one](https://gist.github.com/mshwf/9ef9d5697a6ddaf31447f590c9a790b6)*, If it's possible to add the article then I need some help to ensure it meet the quality of docs content.
Thanks in advance.

(*) video: https://www.youtube.com/watch?v=9FZhLZuWBsE&feature=youtu.be

**Comment by conceptdev** [comment]
Thanks for the offer @mshwf - it is a good suggestion. Perhaps you could propose an outline of the doc you would write (ie. maybe the first paragraph, and the headings for each section)? We'll take a look at the example you linked, and see if we have any other questions. 
If we decide to proceed, we'll let you know and create a branch that you can fork from to start work.

**Comment by mshwf** [comment]
Thanks for accepting my proposal @conceptdev , can I post the outline in this thread?

**Comment by conceptdev** [comment]
Yes, please do!

**Comment by mshwf** [comment]
# Creating Custom Controls #
### Introduction: ###

The process of building UI elements requires, at some point, some customizations to give the unique feel and look to the application and to extend the functionality of existing controls. Whether the customization is just overriding the default TextColor of the Entry or creating a brand new control with new a look and behavior, building custom controls can provide the simple solution to achieve that without the need of custom renderers.

**The process for creating custom controls is as follows:**

1. Create a subclass of the view you want to extend or modify its functionality.
2. Alter the functionality of the subclass by overriding the base class’s bindable properties or create new bindable properties that will interact with the user actions
3. Process the inputs of the user interactions through the propertyChanged delegate of the newly added bindable properties.


# TogglesBar #
### Introduction: ###
In this article we are going to create a group of toggleable buttons (TogglesBar), to show some options that the user can choose from (for example a filtering mechanism) or a light-weight tabbed control..etc, see the screenshot below (should look the same on iOS):
[screenshot (on Android)]

Before we create the TogglesBar group, we need first to create the single ToggleButton control, the steps are as follows:
(the 3 steps above)

[Explaining each step]

**Comment by conceptdev** [comment]
@mshwf sorry for the delay - //build made it a busy week. In the next day or so I'll create a new branch and file (with metadata). You can then fork and work on that branch, and ultimately create a PR for us to review. I'll comment again when it's ready.

**Comment by mshwf** [comment]
Welcome back @conceptdev, I'm ready at anytime

**Comment by conceptdev** [comment]
@mshwf I have created a branch `custom-controls-contrib` with a file for this content **docs/xamarin-forms/user-interface/custom-controls.md**

Please fork from there and edit that file for your submission. You can read about contribution guidelines for docs.microsoft.com here:

https://docs.microsoft.com/en-us/contribute/

If you need to add images, create a directory **custom-controls-images/** and place them there. All filenames must be lowercase. See other pages for examples.

I'd also recommend following some of our other docs for style and structure.

https://docs.microsoft.com/en-us/xamarin/xamarin-forms/app-fundamentals/custom-renderer/entry

We can use this thread for any questions or discussion. Please tag me when you create a PR. Unfortunately we do not have a publicly-available staging server for you to view the final rendering; but we will do a visual check and edit pass before final publication.

Thanks!

**Comment by mshwf** [comment]
It seems that readers of the article will need a sample project, can it be added to xamarin-forms-sample repository?

**Comment by mshwf** [comment]
I have another query:
The control (ToggleBar) uses another custom control (ToggleButton), so I must explain the creation of the ToggleButton first (and I almost finished this [task](https://github.com/mshwf/xamarin-docs/blob/custom-controls-contrib/docs/xamarin-forms/user-interface/custom-controls.md), I will revise it)

Should I proceed to a next part (creating the ToggleBar control) or the ToggleButton covers the topic sufficiently? I can proceed to the next part if you see so, or if there're missing topics from the first part.

**Comment by mshwf** [comment]
Hi, @conceptdev , I just updated the article:
- I concentrated on the ToggleButton control (the child control of the ToggleBar control), to not repeat the same details, and mentioned that:

> Every button inside the toggle bar control is a custom control by itself. This documentation will guide you through creating the Toggle button control and the same concepts can be leveraged in the Toggle bar control (see the complete sample), the steps are as follows:

- I will provide a sample that contains the complete control.
- I still need an answer to my last comments.
- if the quality of the article doesn't match the quality of docs, I'll understand that and work on it.

Thanks!

**Comment by conceptdev** [comment]
Thanks for your work on this! We're checking it out today and will provide more feedback shortly.

**Comment by mshwf** [comment]
@conceptdev I just updated the article; wrote further explanation on the `CustomPropertyChanged` method and renamed the `box` field to `underLine` to describe its role in the control.

**Comment by mshwf** [comment]
I added a note on different types of bindable properties in custom controls:

> [!NOTE] There are two types of custom bindable properties:
> 
> Bindable properties that are passed down to the built-in bindable properties of child elements, like Text bindable property of the ToggleButton custom control, that is passed down to the Text bindable property of the Label control.
> Bindable properties that are specific to the custom control itself and not owned exclusively by any of the child elements, like the IsSelected bindable property, the more behavioral customization required to the custom control, the more of these bindable properties are needed.

 I noticed these two different types from all custom controls I created.

**Comment by mshwf** [comment]
I've added the C# implementation of consuming the ToggleButton control, along with the XAML implementation.  I followed the "Customizing an Entry" article in this part.

**Comment by mshwf** [comment]
@conceptdev I've made my last commit to the article: added the [Summary](https://github.com/mshwf/xamarin-docs/blob/custom-controls-contrib/docs/xamarin-forms/user-interface/custom-controls.md#summary) and [Related Links](https://github.com/mshwf/xamarin-docs/blob/custom-controls-contrib/docs/xamarin-forms/user-interface/custom-controls.md#related-links) sections. During the past days, I've made commits that added explanatory sections, fixed some grammar errors, and making sure the article has the same tune as the docs articles. Please make sure you are checking the last version. Thanks!

**Comment by mshwf** [comment]
Hi @conceptdev, I rewrote the article, here's the [new one](https://github.com/mshwf/xamarin-docs/blob/mshwf-custom-controls-noPropertyChanged/docs/xamarin-forms/user-interface/custom-controls.md), I'm no longer using the `propetyChanged` delegate for initializing the control, instead I used data binding. also I used `ContentView` instead of `StackLayout` as the base class for the custom control. There are some minor changes also.
